### ISSUE TITLE
Adrenaline 1.0.0 feedback

### ISSUE THREAD

**Issue by gyzerok** [gyzerok/adrenaline#98]
Today I've published 1.0.0-rc1 to NPM and would like to here any feedback about new API. Feel free to ask questions and post your thoughts here.

**Comment by vincentfretin** [comment]
Hi @gyzerok 
I started using it on React Native app with a GraphQL endpoint implemented with graphene in Python.
That is great lib, so simple. One missing point is handling promise rejection, you need to call .then(parseJSON).catch(e => console.log(e)) at least, or you don't have any feedback of what's going on in case the server is not responding. A configurable onPromiseRejected callback may be nice to have, so you can show to the user that the request had a problem and add a Retry button.

In your parseJSON function, if it's res.status == 400 (Bad request, an error in your query), I would really like to have the res json parsed, because it contains precious errors to see!

code not tested, it may look like this:
```
function parseJSON(res) {
  if (res.status !== 200) {
    const onError = this.props.onError;
    if (onError) {
      if (res.status == 400) {
        res.json().then(json => onError(json));
      } else {
        onError(res)
    } else {
      throw new Error('Invalid request.');
    }
  }

  return res.json().then(json => json.data);
}
```

Then you can specify a onError(json) function to do a console.log(json.errors ? json.errors : json) when you're developing your app.

**Comment by vincentfretin** [comment]
It seems you already do a .catch(err => { console.err(err); })
in the container component for the query function, great, but you didn't do it for the mutate function.
So I guess it should be there to check for an optional onPromiseRejected prop and call it if defined instead of doing a console.err(err)
What do you think?

**Comment by vincentfretin** [comment]
I'm probably going to use a redux store in my project, so it may be great to add an optional onQueryDone callback to be able to call a redux action to update the redux store instead of doing a setState in the container component.

**Comment by vincentfretin** [comment]
If I understand the code correctly, when you do
```
this.setState({ data: null }, () => {
          this.context.query(query, variables)
            .catch(err => {
              console.err(err); // eslint-disable-line
            })
            .then(data => this.setState({ data }));
```
it probably triggers the render twice, once for each setState. Why not using initialState = {data:null} in constructor and call this.context.query(..) here directly?

If I want to refresh the data (see Pull to refresh functionality in RN http://facebook.github.io/react-native/docs/refreshcontrol.html), I can call this.props.query() in my component to trigger the query a second time. Currently, if I understand the code, it will set again data to null and trigger the query, so my list will probably flicker (be empty some ms, and then render some items).
What it should do is keep the old state during the query and override the state once the new query is performed, be careful to have isFetching to true the second time as well.

**Comment by vincentfretin** [comment]
If I want to implement infinite scrolling with ListView (React Native) or react-list (for the Web), I need to keep all retrieved items in the state. At the end of the first 30 items for example, I would use some redux action like dispatch({type: SET_VARIABLES, first: 30, after: "cursor"}) to trigger a new query with params (first:30, after:"cursor") to fetch the next 30 items. The new results should be appended to the current state.
If I use redux to store the data, I imagine I would use something like this: https://github.com/reactjs/redux/blob/master/examples/real-world/reducers/paginate.js#L33
To do this, I really need to be able to say how the state is mutated (see onQueryDone I proposed in previous comment)

**Comment by gyzerok** [comment]
Hi @vincentfretin! Sorry for such a long time to answer. Gonna give you detailed answer later this day.

**Comment by vincentfretin** [comment]
@gyzerok Don't apologize, it was the weekend after all. ;)

**Comment by gyzerok** [comment]
@vincentfretin I need to think more about how to correctly operate on request error situation.

I don't think `onQueryDone` and `onPromiseRejected` is the right way to handle it. Mostly because I'd like to free decorated component from some non pure stuff and just provide an ability to receive props and make all subsequent decisions around it.

You can achieve onQueryDone behaviour simply with life-cycle hooks
```js
componentDidUpdate(prevProps) {
  const queryIsDone = prevProps.isFetching === true && this.props.isFetching === false;
  if (queryIsDone) {
    // do some stuff
  }
}
```

In case of network or GraphQL errors I need to somehow pass them inside decorated component in a good way. Just need to choose correct way for this.

> It seems you already do a .catch(err => { console.err(err); }) in the container component for the query function, great, but you didn't do it for the mutate function.

Yeah, I do not do it to provide developer ability to decide himself how to handle mutation error.
```js
this.props.mutate(options)
  .then(() => /* do something on success if necessary */)
  .catch(err => /* do something on error if necessary */);
```

**Comment by gyzerok** [comment]
Also I've done temporary fix about data blinking here: https://github.com/gyzerok/adrenaline/commit/81c49f3aac20b5750731846546417dc55a83e2d0

**Comment by gyzerok** [comment]
Maybe you have an idea about the right way of passing errors down as props instead of using callbacks?

**Comment by vincentfretin** [comment]
For queryIsDone, my thought was to call queryIsDone(data) so I can update a store. Actually I may have a better idea. We could add an optional dataReducer prop to the container component and modify the code from this:

      query = () => {
        const { query } = specs;
        const variables = mapPropsToVariables(this.props);

        this.setState({ isFetching: true }, () => {
          this.context.query(query, variables)
            .catch(err => {
              console.err(err); // eslint-disable-line
            })
            .then(data => this.setState({ data, isFetching: false }));
        });
      }

to:

      query = () => {
        const { query } = specs;
        const variables = mapPropsToVariables(this.props);
        dataReducer = this.props.dataReducer | (state, data) => ({ data, isFetching: false })

        this.setState({ isFetching: true }, () => {
          this.context.query(query, variables)
            .catch(err => {
              console.err(err); // eslint-disable-line
            })
            .then(data => this.setState(dataReducer(this.state, data)));
        });
      }

And in my project for an infinite scroll functionality where I want to keep the old data I can do this:

    <container dataReducer={(state, data) => ({ data: [...state.data, data], isFetching: false })}>

Actually it's a more complicated reducer I need to write because this is data.edges I want to append, but you get this idea.

**Comment by gyzerok** [comment]
@vincentfretin sounds very interesting! Let me think about it for some time :)

**Comment by vincentfretin** [comment]
For the errors, you're right, don't use a callback and instead pass a hasErrors and errors props to the decorated component. With this, the decorated component can show a message and a Retry button which just call this.props.query()
hasErrors is a boolean you set to true when you catch the promise, errors is the json.errors.
It's great for 400 bad request, it's great for 502 Service unavailable too, but for the 502, we maybe can add a timeout and retryDelays support in network layer like Relay does:

    Relay.injectNetworkLayer(
      new Relay.DefaultNetworkLayer('http://example.com/graphql', {
        fetchTimeout: 30000,   // Timeout after 30s.
        retryDelays: [5000, 10000],   // Retry after a 5s delay, and then a second time at 10s.
      })
    );

and of course if the auto retries failed, you have hasErrors to true.

**Comment by vincentfretin** [comment]
For the dataReducer part, it's indeed more complicated than that.
To have an idea of what I'm talking about, here is my query:

```javascript
 export default container({                                                      
  query: `                                                                      
     query {                                                                    
      events(                                                   
         first: 30,                                                             
         after: "",                                                                                                                   
         categories: ["Theater","Concert"]) {                                            
             pageInfo{                                                          
               endCursor,                                                       
               hasNextPage,                                                     
             },                                                                 
             edges{                                                             
                node{                                                           
                    ${Event.getFragment('event')}                               
                }                                                               
            }                                                                   
        }                                                                       
    }                                                                           
  `,                                                                            
})(EventList)
```
I actually want to append to data.edges only if the "after" variable changed (I call the query again with the last endCursor in the after variable)
I want a fresh state if other variables changed, say categories here.
Well I really need to do this with redux actions and redux store.

**Comment by gyzerok** [comment]
Btw, if you are going with this Relay-like queries why don't you simply use Relay?

**Comment by vincentfretin** [comment]
As far as I read, Relay doesn't support an offline cache yet. I plan to use [redux-persist](https://github.com/rt2zz/redux-persist) to save the data in the AsyncStorage. I just saw the open sourced f8app did just that. My team already did some projects with redux, so are more confortable with it.
Relay does maintain the whole data.edges for you, I didn't understand it until I reread [this article](https://m.alphasights.com/how-to-do-effortless-pagination-with-relay-connections-and-graphql-ruby-a534ffaf5cbf) which use react-list I was talking about in a previous comment.
Maybe I will end up using Relay with Redux, I don't know yet.

**Comment by vincentfretin** [comment]
@gyzerok I finally did my own container implementation like you can see in https://github.com/gyzerok/adrenaline/issues/110#issuecomment-225122888
